pkg/quota/image/usage.go

```diff
@@ -148,133 +152,117 @@ func (c *GenericImageStreamUsageComputer) GetProjectImagesUsageIncrement(
-			}
-
-			if processedImages.Has(imageName) {
-				glog.V(4).Infof("Skipping image %q - already processed", imageName)
 				continue
 			}
-			processedImages.Insert(imageName)
 
 			if value, ok := img.Annotations[imageapi.ManagedByOpenShiftAnnotation]; !ok || value != "true" {
-				glog.V(5).Infof("Image %q with DockerImageReference %q belongs to an external registry - skipping", img.Name, img.DockerImageReference)
+				glog.V(5).Infof("image %q with DockerImageReference %q belongs to an external registry - skipping", img.Name, img.DockerImageReference)
 				continue
 			}
 
-			cacheInternalRegistryName(img.DockerImageReference)
+			c.cacheInternalRegistryName(img.DockerImageReference)
 
-			if err := handler(tag, history.Items[i].DockerImageReference, img); err != nil {
-				return err
-			}
+			res[img.Name] = img
 		}
 	}
 
-	if c.processSpec {
-		return c.processImageStreamSpecImages(is, processedImages, handler)
-	}
-
-	return nil
+	return res
 }
 
-// processImageStreamSpecImages is a utility method that calls a given handler on every image of the given image
-// stream that belongs to internal registry. It process image stream's spec only.
-func (c *GenericImageStreamUsageComputer) processImageStreamSpecImages(
-	is *imageapi.ImageStream,
-	processedImages sets.String,
-	handler func(tag string, dockerImageReference string, image *imageapi.Image) error,
-) error {
+// gatherImagesFromImageStreamSpec is a utility method that collects all internally managed images found in a
+// spec of the given image stream
+func (c *GenericImageStreamUsageComputer) gatherImagesFromImageStreamSpec(is *imageapi.ImageStream) ImageByName {
+	res := make(ImageByName)
+
 	for tag, tagRef := range is.Spec.Tags {
 		if tagRef.From == nil {
 			continue
 		}
 
 		ref, err := c.getImageReferenceForObjectReference(is.Namespace, tagRef.From)
 		if err != nil {
-			glog.V(4).Infof("Could not process object reference: %v", err)
+			glog.V(4).Infof("could not process object reference: %v", err)
 			continue
 		}
 
-		img, exists := c.imageCache[ref.ID]
-		if !exists {
-			if c.fetchImagesFromImageStream || len(ref.ID) == 0 {
-				if len(ref.ID) > 0 {
-					img, err = c.getImageStreamImage(ref.Namespace, is.Name+"@"+ref.ID)
-					if err != nil {
-						glog.Errorf("Failed to get image stream image %s/%s@%s: %v", ref.Namespace, ref.Name, ref.ID, err)
-						continue
-					}
-
-				} else {
-					tag = ref.Tag
-					if len(tag) == 0 {
-						tag = "latest"
-					}
-					ist, err2 := c.osClient.ImageStreamTags(ref.Namespace).Get(ref.Name, tag)
-					if err2 != nil {
-						glog.Errorf("Failed to get image stream tag %s/%s:%s: %v", ref.Namespace, ref.Name, tag, err2)
-						continue
-					}
-					img = &ist.Image
-					c.imageCache[ref.ID] = img
-				}
+		var img *imageapi.Image
 
-			} else {
-				img, err = c.getImage(ref.ID)
-				if err != nil {
-					glog.Errorf("Failed to get image %s: %v", ref.ID, err)
+		imageObject, exists, err := c.imageCache.GetByKey(ref.ID)
```

Comment:
Can you quickly summarize why you have this cache?  We shouldn't need this anymore because we're only counting unique references.